src/persist-client/src/impl/state.rs

```diff
@@ -0,0 +1,455 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use anyhow::anyhow;
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use mz_persist::location::SeqNo;
+use mz_persist_types::{Codec, Codec64};
+use serde::{Deserialize, Serialize};
+use timely::progress::{Antichain, Timestamp};
+use timely::PartialOrder;
+
+use crate::error::{InvalidUsage, NoOp};
+use crate::read::ReaderId;
+use crate::write::WriterId;
+use crate::Id;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ReadCapability<T> {
+    pub seqno: SeqNo,
+    pub since: Antichain<T>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct WriteCapability<T> {
+    pub upper: Antichain<T>,
+}
+
+// TODO: Document invariants.
+#[derive(Debug)]
+pub struct State<K, V, T, D> {
+    id: Id,
+
+    writers: HashMap<WriterId, WriteCapability<T>>,
+    readers: HashMap<ReaderId, ReadCapability<T>>,
+
+    since: Antichain<T>,
+    trace: Vec<(Vec<String>, Description<T>)>,
+
+    _phantom: PhantomData<(K, V, D)>,
+}
+
+// Impl Clone regardless of the type params.
+impl<K, V, T: Clone, D> Clone for State<K, V, T, D> {
+    fn clone(&self) -> Self {
+        Self {
+            id: self.id.clone(),
+            writers: self.writers.clone(),
+            readers: self.readers.clone(),
+            since: self.since.clone(),
+            trace: self.trace.clone(),
+            _phantom: self._phantom.clone(),
+        }
+    }
+}
+
+impl<K, V, T, D> State<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Codec64,
+{
+    pub fn new(id: Id) -> Self {
+        State {
+            id,
+            writers: HashMap::new(),
+            readers: HashMap::new(),
+            since: Antichain::from_elem(T::minimum()),
+            trace: Vec::new(),
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn id(&self) -> Id {
+        self.id
+    }
+
+    pub fn register(
+        &mut self,
+        seqno: SeqNo,
+        writer_id: &WriterId,
+        reader_id: &ReaderId,
+    ) -> (WriteCapability<T>, ReadCapability<T>) {
+        let write_cap = WriteCapability {
+            upper: self.upper(),
+        };
+        self.writers.insert(writer_id.clone(), write_cap.clone());
+        let read_cap = ReadCapability {
+            seqno,
+            since: self.since.clone(),
+        };
+        self.readers.insert(reader_id.clone(), read_cap.clone());
+        (write_cap, read_cap)
+    }
+
+    pub fn clone_reader(&mut self, seqno: SeqNo, reader_id: &ReaderId) -> ReadCapability<T> {
+        let read_cap = ReadCapability {
+            seqno,
+            since: self.since.clone(),
+        };
+        self.readers.insert(reader_id.clone(), read_cap.clone());
+        read_cap
+    }
+
+    pub fn append(
```

Comment:
quite sus. I got sick of trying to tweak it before we really nailed down our semantics for append and compare_and_append though